include/GafferTest/ContextTest.h

```diff
@@ -37,13 +37,117 @@
+	const V *bPointer = baseContext->getIfExists<V>( "b" );
+
+	{
+		// Scope an editable copy of the context
+		Gaffer::Context::EditableScope scope( baseContext.get() );
+
+		const Gaffer::Context *currentContext = Gaffer::Context::current();
+		GAFFERTEST_ASSERT( currentContext != baseContext );
+
+		// The editable copy should be identical to the original,
+		// and the original should be unchanged.
+		GAFFERTEST_ASSERT( baseContext->get<V>( "a" ) == aVal );
+		GAFFERTEST_ASSERT( baseContext->get<V>( "b" ) == bVal );
+		GAFFERTEST_ASSERT( currentContext->get<V>( "a" ) == aVal );
+		GAFFERTEST_ASSERT( currentContext->get<V>( "b" ) == bVal );
+		GAFFERTEST_ASSERT( currentContext->hash() == baseContext->hash() );
+
+		// The copy should even be referencing the exact same data
+		// as the original.
+		GAFFERTEST_ASSERT( baseContext->getIfExists<V>( "a" ) == aPointer );
+		GAFFERTEST_ASSERT( baseContext->getIfExists<V>( "b" ) == bPointer );
+		GAFFERTEST_ASSERT( currentContext->getIfExists<V>( "a" ) == aPointer );
+		GAFFERTEST_ASSERT( currentContext->getIfExists<V>( "b" ) == bPointer );
+
+		// Editing the copy shouldn't affect the original
+		scope.set( "c", &aVal );
+		GAFFERTEST_ASSERT( baseContext->getIfExists<V>( "c" ) == nullptr );
+		GAFFERTEST_ASSERT( currentContext->get<V>( "c" ) == aVal );
+		GAFFERTEST_ASSERT( currentContext->hash() != baseContext->hash() );
+
+		// Even if we're editing a variable that exists in
+		// the original.
+		scope.set( "a", &bVal );
+		GAFFERTEST_ASSERT( baseContext->get<V>( "a" ) == aVal );
+		GAFFERTEST_ASSERT( currentContext->get<V>( "a" ) == bVal );
+
+		// And we should be able to remove a variable from the
+		// copy without affecting the original too.
+		scope.remove( "b" );
+		GAFFERTEST_ASSERT( baseContext->get<V>( "b" ) == bVal );
+		GAFFERTEST_ASSERT( currentContext->getIfExists<V>( "b" ) == nullptr );
+
+		// And none of the edits should have affected the original
+		// data at all.
+		GAFFERTEST_ASSERT( baseContext->getIfExists<V>( "a" ) == aPointer );
+		GAFFERTEST_ASSERT( baseContext->getIfExists<V>( "b" ) == bPointer );
+
+		// Test setAllocated with Data
+		scope.setAllocated( "a", aData.get() );
+		scope.setAllocated( "b", aData.get() );
+		GAFFERTEST_ASSERT( currentContext->get<V>( "a" ) == aVal );
+		GAFFERTEST_ASSERT( currentContext->get<V>( "b" ) == aVal );
+		GAFFERTEST_ASSERT( currentContext->get( "a" )->isEqualTo( aData.get() ) );
+		GAFFERTEST_ASSERT( currentContext->get( "b" )->isEqualTo( aData.get() ) );
+
+		// And setAllocated with a direct data
+		scope.setAllocated( "b", bVal );
+		GAFFERTEST_ASSERT( currentContext->get<V>( "a" ) == aVal );
+		GAFFERTEST_ASSERT( currentContext->get<V>( "b" ) == bVal );
+
+		// Test getting as a generic Data - this should work where set as Data, or directly from a value
+		GAFFERTEST_ASSERT( currentContext->get( "a" )->isEqualTo( aData.get() ) );
+		GAFFERTEST_ASSERT( currentContext->get( "b" )->isEqualTo( bData.get() ) );
+	}
```

Comment:
One thing that I think is missing here is a check that we have the same `variableHash()` and `hash()` whether or not we set via `Data *`, `T &` or `T *`. Would be good to add that.

Reply:
Added in 570e8cf